runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1266,6 +1508,91 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+   // is available about the class of the array, or the operand is not
+   // even definitely known to be an array, VP has to assume that it might
+   // have a component type that is a value type
+   //
+   else if (!(arrayConstraint && arrayConstraint->getClass()
+              && arrayConstraint->getClassType()->isArray() == TR_yes))
+      {
+      isValueType = TR_maybe;
+      }
+   else
+      {
+      TR_OpaqueClassBlock *arrayComponentClass = fe()->getComponentClassFromArrayClass(arrayConstraint->getClass());
+
+      // Cases to consider:
+      //
+      //   - Is no information available about the component type of the array?
+      //     If not, assume it might be a value type.
+      //   - Is the component type definitely a value type?
+      //   - Is the component type an array class (i.e., is this an array of
+      //     arrays)?  If so, it's definitely not a value type.
+      //   - Is the component type either an abstract class or an interface
+      //     (i.e., not a concrete class)?  If so, it might be a value type.
+      //   - Is the array an array of java/lang/Object?  See below.
+      //
+      // Future refinements:
+      //   The JVM will have classes whose instances must be identity
+      //   objects implement the java/lang/IdentityObject interface, including
+      //   array classes.  Once that happens, the check for isClassArray can
+      //   be replaced with a test of whether the component class implements
+      //   IdentityObject, and the test of !isConcreteClass can be refined
+      //   to distinguish abstract classes that implement IdentityObject from
+      //   those that do not, as well as interfaces that extend IdentityObject.
+      //
+      //   Another potential improvement would be to look for fixed class
+      //   arrays of an interface type
+      //
+      if (!arrayComponentClass)
+         {
+         isValueType = TR_maybe;
+         }
+      else if (TR::Compiler->cls.isValueTypeClass(arrayComponentClass))
+         {
+         isValueType = TR_yes;
+         }
+      else if (TR::Compiler->cls.isClassArray(comp(), arrayComponentClass))
+         {
+         isValueType = TR_no;
+         }
+      else if (!TR::Compiler->cls.isConcreteClass(comp(), arrayComponentClass))
+         {
+         isValueType = TR_maybe;
+         }
+      else
+         {
+         int32_t len;
+         const char *sig = arrayConstraint->getClassSignature(len);
+
+         // If the array's class is a fixed array of java/lang/Object, the
+         // component type is not a value type (though it can still hold
+         // references to instances of value types).
+         if (sig && arrayConstraint->isFixedClass() && sig[0] == '[' && len == 19
+             && !strncmp(sig, "[Ljava/lang/Object;", 19))
+            {
+            isValueType = TR_no;
+            }
+         else
+            {
+            isValueType = TR_maybe;
+            }
```

Comment:
At some point I made a fix to the condition used for arrays of `java/lang/Object`, and ended up making things pessimistic for arrays of other kinds of concrete classes - they'll end up returning `TR_maybe` rather than `TR_no`.  I think this last condition should look something like this:

```
   // If the array is an array of java/lang/Object, and it is fixed to
   // that type, the component type is not a value type (though it
   // can still hold references to instances of value types).  If it is
   // an array of java/lang/Object, but not fixed to that type, the
   // component type could sometimes be a value type.  If it is
   // not an array of java/lang/Object, we know the component
   // must be a concrete class that is not a value type.
   if (sig && sig[0] == '[' && len == 19
       && !strncmp(sig, "[Ljava/lang/Object;", 19))
      {
      isValueType = (arrayConstraint->isFixedClass()) ? TR_no : TR_maybe;
      }
   else
      {
      isValueType = TR_no;
      }
```

Reply:
Addressed this comment in commit cfd28b6